### PR TITLE
curl: add variables to --write-out to allow it used as custom error message

### LINKED ISSUE #6199: Parallel request failures don't include url in error message
* Linux 5.9.0

When making use of the new parallel feature it it appears that any request failures don't include the url from which the error came from.  For example I have seen the following:

```
curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
curl: (22) The requested URL returned error: 502 
```

Due to the lack of a url in the message itself, it's hard to know where these errors are from to take any corrective action.

### `curl` version
```
curl 7.74.0-DEV (x86_64-pc-linux-gnu) libcurl/7.74.0-DEV OpenSSL/1.1.1h zlib/1.2.11 zstd/1.4.5 libidn2/2.3.0 libpsl/0.21.1 (+libidn2/2.3.0) nghttp2/1.41.0 librtmp/2.3
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher http https imap imaps mqtt pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets zstd
```

### PR DESCRIPTION
**Pull request by bagder**
In particular, these ones can help a user to create its own error message when one or more transfers fail.
    
 onerror - lets a user only show the rest on non-zero exit codes
    
 url - the input URL used for this transfer
    
 urlnum - the numerical URL counter (0 indexed) for this transfer
    
 exitcode - the numerical exit code for the transfer
    
 errormsg - obvious
    
Reported-by: @Earnestly
Fixes #6199

### CODE REVIEW

**Comment by jay** [comment]
Does it have to alter the current output? I think that's unnecessary. I take his point but is there something we can add for just those parallel users without affecting everyone else? What about a --show-url-on-error that defaults to true for parallel and false otherwise?

**Comment by bagder** [comment]
> I take his point but is there something we can add for just those parallel users without affecting everyone else?

Yes we could, but then I tried a little bit with command lines like "curl fail-url working-url fail-url" and it struck me that for these cases a URL in there could certainly help as well.

> What about a -show-url-on-error that defaults to true for parallel and false otherwise?

That could work too!

**Comment by bagder** [comment]
Extending that thought: maybe provide a way to let the user select what to put in the error message, similar to `--write-out`, with two different default sets, one for serial and one for parallel...

**Comment by jay** [comment]
Couldn't they just use write-out for that? I don't see much advantage in duplicating it.

**Comment by bagder** [comment]
It could indeed, if we extend the write-out functionality with "request URL" and "input URL index" (and perhaps make it possible to only show if curl returns non-zero)...

**Comment by bagder** [comment]
Still experimenting.
~~~
$ curl -Z 'localmoo/[1-3]' localmoo
...
curl: (6) Could not resolve: localmoo (Domain name not found) [URL 1: localmoo/1]
curl: (6) Could not resolve: localmoo (Domain name not found) [URL 1: localmoo/2]
curl: (6) Could not resolve: localmoo (Domain name not found) [URL 2: localmoo]
curl: (6) Could not resolve: localmoo (Domain name not found) [URL 1: localmoo/3]

**Comment by jay** [comment]
what about like this

~~~
curl: (6) Could not resolve: localmoo (Domain name not found)
URL: localmoo/1

curl: (6) Could not resolve: localmoo (Domain name not found)
URL: localmoo/2

curl: (6) Could not resolve: localmoo (Domain name not found)
URL: localmoo

curl: (6) Could not resolve: localmoo (Domain name not found)
URL: localmoo/3
~~~

**Comment by bagder** [comment]
Well, the number could also be valuable when the same URL is done multiple times in the command line and a subset fails. Like this when the middle transfer fails and the other work:
~~~
$ curl -Z localhost -o /tmp/okay localhost -o /tm -sS localhost -o /tmp/okay2
curl: (23) Failure writing output to destination [URL 2: localhost]

**Comment by jzakrzewski** [comment]
Hmmm... you're indexing from 1? I'm going to get bitten by that just about every time ;)

**Comment by bagder** [comment]
Haha, I actually struggled with that for a while myself to decide. I was maybe hoping that humans start at 1... =)

**Comment by jzakrzewski** [comment]
I'm afraid I cannot be counted as a human anymore.
In any case I won't argue - the curl-tool users are probably still humans ;)

**Comment by bagder** [comment]
@jay's idea tickled my imagination so here I've expanded `--write-out` here with a set of variables to allow users to use it instead of the provided error message:

    onerror - lets a user only show the rest on non-zero exit codes
    
    url - the input URL used for this transfer
    
    urlnum - the numerical URL counter (0 indexed) for this transfer
    
    exitcode - the numerical exit code for the transfer
    
    errormsg - obvious

## Example
~~~
$ curl localhost -o /tmp/okay localhost -o /tm -s localhost -o /tmp/okay2 --write-out "%{onerror}curl: (%{exitcode}) %{errormsg} %{url} %{urlnum}\n"

**Comment by bagder** [comment]
With this `--write-out` change, I can consider backpedaling on the change of the error message and leave it as-is and instead advocate that users craft their own if they need and want more and other info. Would others agree with that?

File: src/tool_operate.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ struct per_transfer {
   struct metalinkfile *mlfile;
   struct metalink_resource *mlres;
   char *this_url;
+  int urlnum; /* the index of the given URL */
```
Comment (emilengler) [inline src/tool_operate.h:44]: Does this need to be an int and signed?

Comment (bagder) [reply src/tool_operate.h:44]: Good remark. I'll at least make it unsigned. I would think that a 32 bit counter is large enough?

**Comment by bagder** [comment]
This PR is now leaving the error message unmodified and instead makes it possible to custom `--write-out` to use as a replacement error message.

**Comment by Earnestly** [comment]
How will `write-out` interact with `-K` configs and whether or not the error message is directed to standard error or standard out?

**Comment by bagder** [comment]
-K is just a (different) way to get the options set, it doesn't particularly affect errors or errors messages.

The `--write-out` message can be delivered to stderr with the use of `%{stderr}`.

**Comment by Earnestly** [comment]
If I understand this correctly then I might get the more useful error messages by including the following?

```
generated-config | curl --create-dirs -sSfKZK - --write-out "%{onerror}%{stderr}curl: %{url}: (%{exitcode}) %{errormsg}\n"
```

**Comment by bagder** [comment]
Exactly!

**Comment by Earnestly** [comment]
That seems fair to me, thank you for looking into this.

**Comment by Earnestly** [comment]
The reason why I use `${filename_effective}` with `-Z` is because output is interleaved.  By creating files with `output /tmp/..` for each url I can then loop over them to recover the output.

Because of this though it means I can't use `write-out` to display the more useful error messages.

**Comment by bagder** [comment]
You *could* do something like this:
~~~
    printf 'write-out %%{filename_effective} %%{onerror}: %%{errromsg}\\n\n'

**Comment by Earnestly** [comment]
Does that construction work with `%{stderr}` sending only the messages after `%{onerror}` to standard error while sending `%{filename_effective}` to standard output?

**Comment by bagder** [comment]
Sure, you can change what "stream" to use mid-string. Send the error part to stderr:
~~~
    printf 'write-out %%{filename_effective} %%{onerror}%%{stderr}: %%{errromsg}\\n\n'

**Comment by Earnestly** [comment]
Oh if that's the case then this approach could work well enough for my uses.

**Pull request closed without merging**